### PR TITLE
refactor(gateway): perform DNS resolution of resources in eventloop

### PR DESCRIPTION
**Pull request by thomaseizinger**
With #3391, constructing a new tunnel will no longer be `async` which makes DNS resolution the only `async` component of  `set_peer_connection_request`. In general, adding resources as part of setting up a connection is a duplicated of the logic within `allow_access`.

We solve both of these problems at once by moving the DNS resolution out of `connlib` into the `gateway` binary and perform it as part of the eventloop during a connection setup.

### CODE REVIEW

File: rust/connlib/shared/src/messages.rs [hunk 1]
```diff
@@ -131,8 +131,8 @@ impl Eq for RequestConnection {}
 
 #[derive(Debug, Deserialize, Serialize, Clone, PartialEq, Eq)]
 #[serde(tag = "type", rename_all = "snake_case")]
-pub enum ResourceDescription {
-    Dns(ResourceDescriptionDns),
+pub enum ResourceDescription<TDNS = ResourceDescriptionDns> {
+    Dns(TDNS),
     Cidr(ResourceDescriptionCidr),
 }
```
Comment (thomaseizinger) [inline rust/connlib/shared/src/messages.rs:137]: This is somewhat of a hack to not cause an excessive diff. The problem here is that we are using the types from the wire protocol with the portal all the way throughout connlib. I don't think we should do that but instead, have separate types that describe the protocol. Ideally, `firezone-tunnel` doesn't depend on that and instead offers APIs that are called by the upper layers (gateway and client-implementation). Those layers can then integrate with the portal connection.

Alternatively, we could also integrate the portal connection directly with `firezone-tunnel`. That would drastically reduce the API surface which could honestly be a good thing as the entire control protocol would then be an implementation detail.

For now, I chose a path that keeps the diff minimal.

File: rust/gateway/src/eventloop.rs [hunk 2]
```diff
@@ -156,22 +162,35 @@ impl Eventloop {
...
+                            .await?;
+
+                        let resource = resolve_resource_description(req.resource).await?;
+
+                        let domain_response = tunnel.allow_access(
+                            resource,
+                            req.client.id,
+                            req.expires_at,
+                            req.client.payload.domain,
+                        );
```
Comment (thomaseizinger) [inline rust/gateway/src/eventloop.rs]: This happens here to ensure it is still atomic from the portal's perspective, i.e. we return the `domain_response` as part of the `ConnectionAccepted`. Long term, I think that should be separate and the portal should send us a separate `allow_access` instruction.

File: rust/connlib/shared/src/messages.rs (hunk 1 above)
Comment (conectado) [reply rust/connlib/shared/src/messages.rs:137]: Originally, I was planning on having separated types for the tunnel and the control-protocol but I was waiting to implement it with [this issue](https://github.com/firezone/firezone/issues/2187)

> Alternatively, we could also integrate the portal connection directly with firezone-tunnel. That would drastically reduce the API surface which could honestly be a good thing as the entire control protocol would then be an implementation detail.

At this point we'd just get rid of `Tunnel`, right?

**Review by conectado — APPROVED** [review]
LGTM! :)

**Comment by jamilbk** [comment]
I think this PR is causing a flaky test -- here's the workflow run for the commit this created on `main`:

https://github.com/firezone/firezone/actions/runs/7909913996/job/21591965123

**Comment by ReactorScram** [comment]
https://github.com/firezone/firezone/pull/3654 would revert this one to fix the DNS test in CI

**Comment by ReactorScram** [comment]
Just tested f42aa862a8b in Windows and it does work there. I wonder what's different about Linux. Maybe the CI test is wrong somehow? I'll try to test Linux locally.

Edit: Couldn't replicate it on Debian either. Is there something particular about `test.httpbin.docker.local` or whatever the name was in CI?

**Pull request merged**